### PR TITLE
[WIP] Changed default to cache image layers during build, like docker

### PR DESCRIPTION
**Pull request by fatherlinux**
Signed-off-by: Scott McCarty <smccarty@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
bot, add author to whitelist

**Comment by vrothberg** [comment]
@fatherlinux, thanks for the change! CI breaks as the code needs some formatting (e.g., via `gofmt`).

For Buildah, the default depends on the `BUILDAH_LAYERS` environment variable and unless it's set, Buildah defaults to false. If Docker's default is `true` then I think we should change that in Buildah and vendor in the changes here. @TomSweeneyRedHat @nalind, PTAL.

**Comment by rhatdan** [comment]
@fatherlinux Coding in Go....

I was playing around with this yesterday, attempting to change the defaults of podman to layers and leave buildah default to non-layers.  My goal was to make sure that podman build --help showed the right default.  Sadly I was feeling because of the point where options are handled.

@fatherlinux Should we change the default in both buildah and podman?  Would make the coding easier?

**Comment by fatherlinux** [comment]
@rhatdan up to you. It was my understanding that Buildah could be used by automated builders (aka in Kubernetes), so it might be desirable to have the default caching == false. On the other hand, IMHO, having Podman default to true makes sense for docker compatibility (expected behavior).

**Comment by fatherlinux** [comment]
@vrothberg I am confused. I fixed the formatting with gofmt but some of the tests still fail? It shows that the code is fixed, but in the code highlighter, but still fails...

**Comment by fatherlinux** [comment]
@vrothberg actually, that appears to have been a JavaScript bug with the GitHub interface. It was flipping back and forth between an old version of the test, and the current (in progress). It should pass now :-)

**Comment by fatherlinux** [comment]
@rhatdan also, I updated the Podman man page to reflect the default :-)

**Comment by rhatdan** [comment]
@fatherlinux if you do a `podman build --help | grep layers`
What does it say is the default?

**Comment by TomSweeneyRedHat** [comment]
LGTM assuming happy tests.  
I'm hip with changing the Buildah default for layers if that's the decision.

**Comment by fatherlinux** [comment]
@rhatdan I had to remove the system version of podman (from RHEL repos), modify the man pages again, and do another "make install" - I tested and the man pages are now updated. 

I am not sure what was happening, but something went sideways. Just going to use upstream podman , now that I know how to compile it :-)

**Comment by baude** [comment]
I suspect the help is wrong though ... i suspect it still says false.

File: cmd/podman/build.go [hunk 1]
```diff
@@ -187,6 +187,13 @@ func buildCmd(c *cli.Context) error {
 	}
 	namespaceOptions.AddOrReplace(usernsOption...)
 
+	layers := buildahcli.UseLayers()
```
Comment (mheon) [inline cmd/podman/build.go]: Do we need this? We're unconditionally overwriting it with either c.Bool("layers") or true

**Comment by baude** [comment]
± |master → origin ↑683 U:2 ?:32 ✗| → sudo podman build --help | grep "\-\-layer"
   --layers                                    cache intermediate layers during build. Use BUILDAH_LAYERS environment variable to override. (default false)

So the setting in the help is incorrect.  I think we need to rework the code slightly on the buildah side first.

**Comment by fatherlinux** [comment]
@baude technically, that's right. The BUILDAH_LAYERS environment variable, if not set will default to false. I could rework the wording to make super clear if that is helpful?

**Comment by fatherlinux** [comment]
@baude actually, I just checked the podman manpage for the PR I submitted, here is the full text. It's dead on for Podman:

"--layers

       Cache intermediate images during the build process (Default is true).

       Note: You can also override the default value of layers by setting the BUILDAH_LAYERS environment  variable.  export
       BUILDAH_LAYERS=true"

**Comment by baude** [comment]
but not in the cli when you do --help

File: cmd/podman/build.go (hunk 1 above)
Comment (fatherlinux) [reply cmd/podman/build.go]: @mheon I think we do. I needed an intuitive set of cases for the end user. I tested these cases:
1. podman --layers=true (overrides everything)
2. podman --layers=false (overrides everything)
3. export BUILDAH_LAYERS=true (works, but is overridden by podman --layers=false)
4. export BUILDAH_LAYERS=false (works, but is overridden by podman --layers=true)

c.Bool("layers") gets us the command line option, but then sets the default for podman differently with an else statement. That's why the is.Set logic is there....

Comment (fatherlinux) [reply cmd/podman/build.go]: Also, 5. defaults to true...

**Comment by fatherlinux** [comment]
Ahhh, crud. Will fix.

**Comment by fatherlinux** [comment]
@baude actually, I just tested the build with my patches, and --help does not output anything about --layers:

[root@rhel7 containers]# podman --help
NAME:
   podman - manage pods and images

USAGE:
   podman [global options] command [command options] [arguments...]

VERSION:
   0.8.5-dev

COMMANDS:
     attach           Attach to a running container
     commit           Create new image based on the changed container
     container        Manage Containers
     build            Build an image using instructions from Dockerfiles
     create           Create but do not start a container
     diff             Inspect changes on container's file systems
     exec             Run a process in a running container
     export           Export container's filesystem contents as a tar archive
     history          Show history of a specified image
     image            Manage images
     images           List images in local storage
     import           Import a tarball to create a filesystem image
     info             Display podman system information
     inspect          Displays the configuration of a container or image
     kill             Kill one or more running containers with a specific signal
     load             Load an image from docker archive
     login            Login to a container registry
     logout           Logout of a container registry
     logs             Fetch the logs of a container
     mount            Mount a working container's root filesystem
     pause            Pauses all the processes in one or more containers
     ps               List containers
     pod              Manage pods
     port             List port mappings or a specific mapping for the container
     pull             Pull an image from a registry
     push             Push an image to a specified destination
     restart          Restart one or more containers
     rm               Remove one or more containers
     rmi              Removes one or more images from local storage
     run              Run a command in a new container
     save             Save image to an archive
     search           Search registry for image
     start            Start one or more containers
     stats            Display percentage of CPU, memory, network I/O, block I/O and PIDs for one or more containers
     stop             Stop one or more containers
     tag              Add an additional name to a local image
     top              Display the running processes of a container
     umount, unmount  Unmounts working container's root filesystem
     unpause          Unpause the processes in one or more containers
     version          Display the PODMAN Version Information
     wait             Block on one or more containers
     varlink          Run varlink interface
     help, h          Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --cgroup-manager value            cgroup manager to use (cgroupfs or systemd, default systemd)
   --cni-config-dir value            path of the configuration directory for CNI networks
   --config value, -c value          path of a config file detailing container server configuration options
   --conmon value                    path of the conmon binary
   --cpu-profile value               path for the cpu profiling results
   --log-level value                 log messages above specified level: debug, info, warn, error (default), fatal or panic (default: "error")
   --namespace value                 set the libpod namespace, used to create separate views of the containers and pods on the system
   --root value                      path to the root directory in which data, including images, is stored
   --tmpdir value                    path to the tmp directory
   --runroot value                   path to the 'run directory' where all state information is stored
   --runtime value                   path to the OCI-compatible binary used to run containers, default is /usr/bin/runc
   --storage-driver value, -s value  select which storage driver is used to manage storage of images and containers (default is overlay)
   --storage-opt value               used to pass an option to the storage driver
   --syslog                          output logging information to syslog as well as the console
   --help, -h                        show help
   --version, -v                     print the version

**Comment by rhatdan** [comment]
podman build --help

```
podman build --help | grep -i -- --layers
   --layers                                    cache intermediate layers during build. Use BUILDAH_LAYERS environment variable to override. (default false)
```

**Comment by umohnani8** [comment]
If we are planning on making build with layers default, I think it would be better to get rid of the --layers flag and make that --no-layers? This way we won't have to `podman build --layers=false` and can do `podman build --no-layers`. Same thing for buildah if we are changing the default there too.

**Comment by rhatdan** [comment]
Changeing Buildah is a little more difficult, since it is already supported. BTW I think we would need to make the --layers flag a noop and hidden.

**Comment by umohnani8** [comment]
@rhatdan no-cache just means build with layers but don't use any of the cached images. no-layers would mean build without creating cached images, so what buildah does by default. If we don't want podman to have that feature then no need of --no-layers as you said.

**Comment by rhatdan** [comment]
How much of a speed improvement do we get with --layers=false?

**Comment by umohnani8** [comment]
@rhatdan not sure about the exact numbers, but in the first build --layers=false will be slightly faster but after making a change and in subsequent builds --layers=true might be faster as certain steps will be skipped.

**Comment by rhatdan** [comment]
Yes I know for developer mode that --layers=true is faster.  But for CI/CD or builds within a container then --layers=false should theoretically be faster, since layers are not being committed.

**Comment by baude** [comment]
@rhatdan @umohnani8 what do you guys want to do here? I like the no-cache idea.  Should we make a solution for this week's release?

**Comment by rhatdan** [comment]
I think we leave all options but change --layers=true to be default.

**Comment by rhatdan** [comment]
https://github.com/containers/libpod/pull/1422 fixes this.

**Pull request closed without merging**